File: backend_InMoment/InMoment.Domain/Users/AccountDeletionRequest.cs
using InMoment.Domain.Common;

namespace InMoment.Domain.Users;

public sealed class AccountDeletionRequest : Entity<Guid>
{
    public Guid UserId { get; private set; }

    public string RequestedEmail { get; private set; } = default!;
    public string RequestedUserName { get; private set; } = default!;

    public AccountDeletionRequestStatus Status { get; private set; }

    public string? Note { get; private set; }
    public string? ProcessingNote { get; private set; }

    public DateTime RequestedAtUtc { get; private set; }
    public DateTime UpdatedAtUtc { get; private set; }
    public DateTime? ProcessedAtUtc { get; private set; }
    public Guid? ProcessedByUserId { get; private set; }

    private AccountDeletionRequest() { }

    public static AccountDeletionRequest Create(
        Guid userId,
        string requestedEmail,
        string requestedUserName,
        string? note)
    {
        if (userId == Guid.Empty)
            throw new ValidationException("UserId is required.");

        var normalizedEmail = NormalizeEmail(requestedEmail);
        var normalizedUserName = NormalizeUserName(requestedUserName);
        var normalizedNote = NormalizeNote(note);

        var now = DateTime.UtcNow;

        return new AccountDeletionRequest
        {
            Id = Guid.NewGuid(),
            UserId = userId,
            RequestedEmail = normalizedEmail,
            RequestedUserName = normalizedUserName,
            Status = AccountDeletionRequestStatus.Pending,
            Note = normalizedNote,
            ProcessingNote = null,
            RequestedAtUtc = now,
            UpdatedAtUtc = now,
            ProcessedAtUtc = null,
            ProcessedByUserId = null
        };
    }

    public void MarkInProgress(Guid moderatorUserId, string? processingNote = null)
    {
        EnsureModerator(moderatorUserId);
        EnsureNotTerminal();

        Status = AccountDeletionRequestStatus.InProgress;
        ProcessingNote = NormalizeProcessingNote(processingNote) ?? ProcessingNote;
        UpdatedAtUtc = DateTime.UtcNow;
        ProcessedByUserId = moderatorUserId;
    }

    public void Complete(Guid moderatorUserId, string? processingNote = null)
    {
        EnsureModerator(moderatorUserId);
        EnsureNotTerminal();

        Status = AccountDeletionRequestStatus.Completed;
        ProcessingNote = NormalizeProcessingNote(processingNote) ?? ProcessingNote;
        UpdatedAtUtc = DateTime.UtcNow;
        ProcessedAtUtc = DateTime.UtcNow;
        ProcessedByUserId = moderatorUserId;
    }

    public void Reject(Guid moderatorUserId, string? processingNote = null)
    {
        EnsureModerator(moderatorUserId);
        EnsureNotTerminal();

        Status = AccountDeletionRequestStatus.Rejected;
        ProcessingNote = NormalizeProcessingNote(processingNote) ?? ProcessingNote;
        UpdatedAtUtc = DateTime.UtcNow;
        ProcessedAtUtc = DateTime.UtcNow;
        ProcessedByUserId = moderatorUserId;
    }

    public void Cancel(string? note = null)
    {
        EnsureNotTerminal();

        Status = AccountDeletionRequestStatus.Cancelled;
        Note = NormalizeNote(note) ?? Note;
        UpdatedAtUtc = DateTime.UtcNow;
        ProcessedAtUtc = DateTime.UtcNow;
    }

    private void EnsureNotTerminal()
    {
        if (Status is AccountDeletionRequestStatus.Completed
            or AccountDeletionRequestStatus.Rejected
            or AccountDeletionRequestStatus.Cancelled)
        {
            throw new ValidationException("Request is already finalized.");
        }
    }

    private static void EnsureModerator(Guid moderatorUserId)
    {
        if (moderatorUserId == Guid.Empty)
            throw new ValidationException("Moderator user id is required.");
    }

    private static string NormalizeEmail(string value)
    {
        if (string.IsNullOrWhiteSpace(value))
            throw new ValidationException("RequestedEmail is required.");

        var normalized = value.Trim().ToLowerInvariant();

        if (normalized.Length > 256)
            throw new ValidationException("RequestedEmail is too long.");

        return normalized;
    }

    private static string NormalizeUserName(string value)
    {
        if (string.IsNullOrWhiteSpace(value))
            throw new ValidationException("RequestedUserName is required.");

        var normalized = value.Trim();

        if (normalized.Length > 64)
            throw new ValidationException("RequestedUserName is too long.");

        return normalized;
    }

    private static string? NormalizeNote(string? value)
    {
        if (string.IsNullOrWhiteSpace(value))
            return null;

        var normalized = value.Trim();

        if (normalized.Length > 2000)
            throw new ValidationException("Note is too long.");

        return normalized;
    }

    private static string? NormalizeProcessingNote(string? value)
    {
        if (string.IsNullOrWhiteSpace(value))
            return null;

        var normalized = value.Trim();

        if (normalized.Length > 2000)
            throw new ValidationException("ProcessingNote is too long.");

        return normalized;
    }
}

File: frontend_InMoment/lib/features/account/models/account_deletion_request.dart
class AccountDeletionRequest {
  final String id;
  final String userId;
  final int statusCode;
  final String status;
  final String requestedEmail;
  final String requestedUserName;
  final String? note;
  final String? processingNote;
  final String? processedByUserId;
  final DateTime requestedAtUtc;
  final DateTime updatedAtUtc;
  final DateTime? processedAtUtc;

  const AccountDeletionRequest({
    required this.id,
    required this.userId,
    required this.statusCode,
    required this.status,
    required this.requestedEmail,
    required this.requestedUserName,
    required this.note,
    required this.processingNote,
    required this.processedByUserId,
    required this.requestedAtUtc,
    required this.updatedAtUtc,
    required this.processedAtUtc,
  });

  bool get isActive => statusCode == 1 || statusCode == 2;

  bool get isTerminal => statusCode == 3 || statusCode == 4 || statusCode == 5;

  String get statusLabel {
    switch (statusCode) {
      case 1:
        return 'Ожидает обработки';
      case 2:
        return 'В обработке';
      case 3:
        return 'Удаление выполнено';
      case 4:
        return 'Отклонён';
      case 5:
        return 'Отменён';
      default:
        return status;
    }
  }

  factory AccountDeletionRequest.fromJson(Map<String, dynamic> json) {
    DateTime parseDate(String key) {
      final raw = json[key]?.toString();
      if (raw == null || raw.isEmpty) {
        return DateTime.fromMillisecondsSinceEpoch(0, isUtc: true);
      }

      return DateTime.tryParse(raw)?.toUtc() ??
          DateTime.fromMillisecondsSinceEpoch(0, isUtc: true);
    }

    DateTime? parseNullableDate(String key) {
      final raw = json[key]?.toString();
      if (raw == null || raw.isEmpty) return null;
      return DateTime.tryParse(raw)?.toUtc();
    }

    return AccountDeletionRequest(
      id: (json['id'] ?? '').toString(),
      userId: (json['userId'] ?? '').toString(),
      statusCode: (json['statusCode'] as num?)?.toInt() ?? 0,
      status: (json['status'] ?? '').toString(),
      requestedEmail: (json['requestedEmail'] ?? '').toString(),
      requestedUserName: (json['requestedUserName'] ?? '').toString(),
      note: json['note']?.toString(),
      processingNote: json['processingNote']?.toString(),
      processedByUserId: json['processedByUserId']?.toString(),
      requestedAtUtc: parseDate('requestedAtUtc'),
      updatedAtUtc: parseDate('updatedAtUtc'),
      processedAtUtc: parseNullableDate('processedAtUtc'),
    );
  }
}